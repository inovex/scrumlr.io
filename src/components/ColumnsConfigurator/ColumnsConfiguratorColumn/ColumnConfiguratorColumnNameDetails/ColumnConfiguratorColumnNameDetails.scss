@import "constants/style";

.column-configurator-column-name-details__name-wrapper {
  display: flex;
  flex-direction: column;
  gap: $spacing--xs;

  &:hover {
    // affect both with prio
    .column-configurator-column-name-details__name,
    .column-configurator-column-name-details__inline-description {
      color: $navy--400;
    }
  }
}

// TODO input style wip
.column-configurator-column-name-details__name {
  all: unset;

  align-self: flex-start; // only take as much space as needed
  // extend border to bottom and right
  padding: 0 $spacing--xs $spacing--xs 0;
  max-width: 100%;

  font-size: $text--base;
  font-weight: 600;
  color: $navy--600;
  border-bottom: 2px solid var(--accent-color--light);

  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;

  &:focus {
    color: $navy--600;
    outline: none;
  }

  &::placeholder {
    color: $gray--800;
  }

  // not only when focused, but also when description is being edited
  &--editing {
    border-bottom-style: dashed; // TODO dash spacing
  }
}

.column-configurator-column-name-details__inline-description {
  color: $gray--800;
  font-size: $text--xs;
  font-weight: 500;

  // if in visual feedback mode, show as multiline, else in one line
  &:not(&--visual-feedback) {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }

  &--has-content {
    font-size: 12px;
    font-weight: 500;
    color: $navy--900;
  }
}

.column-configurator-column-name-details__description-wrapper {
  display: flex;
  flex-direction: column;
}

.column-configurator-column-name-details__description-mini-menu {
  margin-left: auto; // couldn't get align-self: flex-end to work so whatever
}

[theme="dark"] {
  .column-configurator-column-name-details__name-wrapper {
    &:hover {
      // affect both with prio
      .column-configurator-column-name-details__name,
      .column-configurator-column-name-details__inline-description {
        color: $gray--000;
      }
    }
  }

  .column-configurator-column-name-details__name {
    color: $gray--000;

    border-bottom-color: var(--accent-color--dark);
  }

  .column-configurator-column-name-details__inline-description {
    color: $gray--700;

    &--has-content {
      font-size: 12px;
      font-weight: 500;
      color: $gray--000;
    }
  }
}
