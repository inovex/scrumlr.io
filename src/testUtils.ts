export type AssertTypeEqual<T, Expected> =
    T extends Expected
        ? (Expected extends T ? true : never)
        : never;