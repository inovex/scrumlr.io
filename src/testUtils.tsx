import {render, RenderOptions} from "@testing-library/react";
import {I18nextProvider} from "react-i18next";
import {FC, PropsWithChildren, ReactElement} from "react";
import {MemoryRouter} from "react-router";
import {Provider} from "react-redux";
import getTestStore from "utils/test/getTestStore";
import i18n from "./i18nTest";

export type AssertTypeEqual<T, Expected> = T extends Expected ? (Expected extends T ? true : never) : never;

const AllTheProvidersWithoutRouter: FC<PropsWithChildren> = ({children}) => (
  <I18nextProvider i18n={i18n}>
    <Provider store={getTestStore()}>{children}</Provider>
  </I18nextProvider>
);

const AllTheProviders: FC<PropsWithChildren> = ({children}) => (
  <MemoryRouter>
    <AllTheProvidersWithoutRouter>{children}</AllTheProvidersWithoutRouter>
  </MemoryRouter>
);

const customRender = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) => render(ui, {wrapper: AllTheProviders, ...options});
const customRenderWithoutRouter = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) =>
  render(ui, {wrapper: AllTheProvidersWithoutRouter, ...options});

export {customRender as render, customRenderWithoutRouter as renderWithoutRouter};
