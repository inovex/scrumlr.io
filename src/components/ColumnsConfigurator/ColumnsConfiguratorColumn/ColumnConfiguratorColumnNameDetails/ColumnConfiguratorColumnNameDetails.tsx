import classNames from "classnames";
import {useTranslation} from "react-i18next";
import {ReactComponent as CheckDoneIcon} from "assets/icons/check-done.svg";
import {ReactComponent as CloseIcon} from "assets/icons/close.svg";
import {TextArea} from "components/TextArea/TextArea";
import {MiniMenu, MiniMenuItem} from "components/MiniMenu/MiniMenu";
import {Dispatch, SetStateAction, useRef, useState, FocusEvent} from "react";
import "./ColumnConfiguratorColumnNameDetails.scss";
import {MAX_COLUMN_DESCRIPTION_LENGTH} from "constants/misc";

export type OpenState = "closed" | "visualFeedback" | "nameFirst" | "descriptionFirst";

export type ColumnConfiguratorColumnNameDetailsProps = {
  className: string;

  name: string;
  description: string;

  // name/description indicate that the editable section should be opened with focus on either the title or description input
  openState: OpenState;
  setOpenState: Dispatch<SetStateAction<OpenState>>;

  updateColumnTitle: (newName: string, newDescription: string) => void;
};

export const ColumnConfiguratorColumnNameDetails = (props: ColumnConfiguratorColumnNameDetailsProps) => {
  const {t} = useTranslation();

  const nameWrapperRef = useRef<HTMLDivElement>(null);

  // temporary state for name and description text as the changes have to be confirmed before applying
  const [name, setName] = useState(props.name);
  const [description, setDescription] = useState(props.description);

  const isEditing = props.openState === "nameFirst" || props.openState === "descriptionFirst";

  const descriptionConfirmMiniMenu: MiniMenuItem[] = [
    {
      className: "mini-menu-item--cancel",
      element: <CloseIcon />,
      label: t("Templates.ColumnsConfiguratorColumn.cancel"),
      onClick(): void {
        props.setOpenState("closed");
        (document.activeElement as HTMLElement)?.blur(); // leave input (or we can keep typing inside it)
      },
    },
    {
      className: "mini-menu-item--save",
      element: <CheckDoneIcon />,
      label: t("Templates.ColumnsConfiguratorColumn.save"),
      onClick(): void {
        props.updateColumnTitle(name, description);
        // show visual feedback for 2s before displaying menu options again
        props.setOpenState("visualFeedback");
        setTimeout(() => {
          props.setOpenState("closed");
        }, 2000);
      },
    },
  ];

  // if we leave the wrapper close, otherwise leave open
  const handleBlurNameWrapperContents = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    const isFocusInsideTitleHeaderWrapper = nameWrapperRef.current?.contains(e.relatedTarget);

    if (!isFocusInsideTitleHeaderWrapper) {
      props.setOpenState("closed");

      // reset name and description to actual
      setName(props.name);
      setDescription(props.description);
    }
  };

  const openDescriptionWithCurrentValue = () => {
    setDescription(props.description);
    props.setOpenState("descriptionFirst");
  };

  return (
    <div className={classNames(props.className, "column-configurator-column-name-details__name-wrapper")} ref={nameWrapperRef}>
      <input
        className={classNames("column-configurator-column-name-details__name", {"column-configurator-column-name-details__name--editing": isEditing})}
        value={name}
        placeholder={t("Templates.ColumnsConfiguratorColumn.namePlaceholder")}
        onInput={(e) => setName(e.currentTarget.value)}
        onFocus={() => props.setOpenState("nameFirst")}
        onBlur={handleBlurNameWrapperContents}
        autoComplete="off"
      />
      {isEditing ? (
        <div className="column-configurator-column-name-details__description-wrapper">
          <TextArea
            className="column-configurator-column-name-details__description-text-area"
            input={description}
            setInput={setDescription}
            placeholder={t("Templates.ColumnsConfiguratorColumn.descriptionPlaceholder")}
            embedded
            fitted
            textDim
            autoFocus={props.openState === "descriptionFirst"}
            onBlur={handleBlurNameWrapperContents}
            maxLength={MAX_COLUMN_DESCRIPTION_LENGTH}
          />
          <MiniMenu className="column-configurator-column-name-details__description-mini-menu" items={descriptionConfirmMiniMenu} small transparent />
        </div>
      ) : (
        <div
          className={classNames("column-configurator-column-name-details__inline-description", {
            "column-configurator-column-name-details__inline-description--visual-feedback": props.openState === "visualFeedback",
          })}
          role="button"
          tabIndex={0}
          onClick={openDescriptionWithCurrentValue}
        >
          {props.description ? props.description : t("Templates.ColumnsConfiguratorColumn.descriptionPlaceholder")}
        </div>
      )}
    </div>
  );
};
