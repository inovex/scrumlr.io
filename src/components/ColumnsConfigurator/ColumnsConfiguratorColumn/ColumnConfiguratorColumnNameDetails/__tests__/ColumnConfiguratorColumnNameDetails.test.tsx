import {ColumnConfiguratorColumnNameDetails, ColumnConfiguratorColumnNameDetailsProps} from "../ColumnConfiguratorColumnNameDetails";
import {render} from "testUtils";
import {fireEvent} from "@testing-library/react";

const renderColumnConfiguratorColumnNameDetails = (override?: Partial<ColumnConfiguratorColumnNameDetailsProps>) => {
  const defaultProps: ColumnConfiguratorColumnNameDetailsProps = {
    className: "column-configurator-column-name-details",
    name: "Default Name",
    description: "Default Description",
    openState: "closed",
    setOpenState: jest.fn(),
    updateColumnTitle: jest.fn(),
  };

  return render(<ColumnConfiguratorColumnNameDetails {...defaultProps} {...override} />);
};

describe("ColumnConfiguratorColumnNameDetails render", () => {
  it("should render correctly (closed)", () => {
    const {container} = renderColumnConfiguratorColumnNameDetails({openState: "closed"});
    expect(container).toMatchSnapshot();
  });

  it("should render correctly (visualFeedback)", () => {
    const {container} = renderColumnConfiguratorColumnNameDetails({openState: "visualFeedback"});
    expect(container).toMatchSnapshot();
  });

  it("should render correctly (nameFirst)", () => {
    const {container} = renderColumnConfiguratorColumnNameDetails({openState: "nameFirst"});
    expect(container).toMatchSnapshot();
  });

  it("should render correctly (descriptionFirst)", () => {
    const {container} = renderColumnConfiguratorColumnNameDetails({openState: "descriptionFirst"});
    expect(container).toMatchSnapshot();
  });

  it("should match title and description", () => {
    const name = "Custom Title";
    const description = "Custom Description";
    const {container} = renderColumnConfiguratorColumnNameDetails({name, description});

    expect(container.querySelector<HTMLInputElement>(".column-configurator-column-name-details__name")).toHaveValue(name);
    expect(container.querySelector<HTMLDivElement>(".column-configurator-column-name-details__inline-description")).toHaveTextContent(description);
  });
});

describe("ColumnConfiguratorColumnNameDetails behaviour", () => {
  it("should close after losing focus", () => {
    const setOpenStateSpy = jest.fn();
    const {container} = renderColumnConfiguratorColumnNameDetails({openState: "nameFirst", setOpenState: setOpenStateSpy});

    const inputElement = container.querySelector<HTMLInputElement>(".column-configurator-column-name-details__name")!;

    fireEvent.focus(inputElement);
    fireEvent.blur(inputElement);

    expect(setOpenStateSpy).toHaveBeenCalledWith("closed");
  });

  it("should set openState to visualFeedback, then to closed upon confirming changes", () => {
    jest.useFakeTimers();
    const setOpenStateSpy = jest.fn();
    const {container} = renderColumnConfiguratorColumnNameDetails({openState: "nameFirst", setOpenState: setOpenStateSpy});

    const saveChangesButtonElement = container.querySelector<HTMLButtonElement>(".mini-menu-item--save")!;
    fireEvent.mouseDown(saveChangesButtonElement);

    expect(setOpenStateSpy).toHaveBeenCalledWith("visualFeedback");

    setTimeout(() => {
      expect(setOpenStateSpy).toHaveBeenCalledWith("closed");
    }, 2000);

    jest.runAllTimers();
  });

  it("should call save changes with new input title", () => {
    const updateColumnTitleSpy = jest.fn();
    const {container} = renderColumnConfiguratorColumnNameDetails({openState: "nameFirst", updateColumnTitle: updateColumnTitleSpy});

    const inputElement = container.querySelector<HTMLInputElement>(".column-configurator-column-name-details__name")!;
    fireEvent.input(inputElement, {target: {value: "Custom Title"}});

    const saveChangesButtonElement = container.querySelector<HTMLButtonElement>(".mini-menu-item--save")!;
    fireEvent.mouseDown(saveChangesButtonElement);

    expect(updateColumnTitleSpy).toHaveBeenCalledWith("Custom Title", expect.anything());
  });
});
