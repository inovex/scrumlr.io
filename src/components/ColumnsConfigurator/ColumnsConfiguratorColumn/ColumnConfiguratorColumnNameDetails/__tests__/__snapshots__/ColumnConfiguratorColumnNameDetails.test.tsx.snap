// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnConfiguratorColumnNameDetails render should render correctly (closed) 1`] = `
<div>
  <div
    class="column-configurator-column-name-details column-configurator-column-name-details__name-wrapper"
  >
    <input
      autocomplete="off"
      class="column-configurator-column-name-details__name"
      placeholder="Title"
      value="Default Name"
    />
    <div
      class="column-configurator-column-name-details__inline-description"
      role="button"
      tabindex="0"
    >
      Default Description
    </div>
  </div>
</div>
`;

exports[`ColumnConfiguratorColumnNameDetails render should render correctly (descriptionFirst) 1`] = `
<div>
  <div
    class="column-configurator-column-name-details column-configurator-column-name-details__name-wrapper"
  >
    <input
      autocomplete="off"
      class="column-configurator-column-name-details__name column-configurator-column-name-details__name--editing"
      placeholder="Title"
      value="Default Name"
    />
    <div
      class="column-configurator-column-name-details__description-wrapper"
    >
      <style>
        .text-area { --text-area-rows: 7 }
      </style>
      <textarea
        class="column-configurator-column-name-details__description-text-area text-area text-area--embedded text-area--fitted text-area--text-dim text-area--border-normal text-area--text-align-left"
        maxlength="1024"
        placeholder="Description (optional)"
      >
        Default Description
      </textarea>
      <div
        class="column-configurator-column-name-details__description-mini-menu mini-menu mini-menu--small mini-menu--transparent"
      >
        <button
          aria-label="Cancel"
          class="mini-menu-item--cancel mini-menu__item mini-menu__item--small"
          data-cy="undefined-item-Cancel"
          data-tooltip-content="Cancel"
          data-tooltip-id="scrumlr-tooltip"
          id="mini-menu-Cancel5"
        >
          <svg>
            close.svg
          </svg>
        </button>
        <button
          aria-label="Save"
          class="mini-menu-item--save mini-menu__item mini-menu__item--small"
          data-cy="undefined-item-Save"
          data-tooltip-content="Save"
          data-tooltip-id="scrumlr-tooltip"
          id="mini-menu-Save6"
        >
          <svg>
            check-done.svg
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ColumnConfiguratorColumnNameDetails render should render correctly (nameFirst) 1`] = `
<div>
  <div
    class="column-configurator-column-name-details column-configurator-column-name-details__name-wrapper"
  >
    <input
      autocomplete="off"
      class="column-configurator-column-name-details__name column-configurator-column-name-details__name--editing"
      placeholder="Title"
      value="Default Name"
    />
    <div
      class="column-configurator-column-name-details__description-wrapper"
    >
      <style>
        .text-area { --text-area-rows: 7 }
      </style>
      <textarea
        class="column-configurator-column-name-details__description-text-area text-area text-area--embedded text-area--fitted text-area--text-dim text-area--border-normal text-area--text-align-left"
        maxlength="1024"
        placeholder="Description (optional)"
      >
        Default Description
      </textarea>
      <div
        class="column-configurator-column-name-details__description-mini-menu mini-menu mini-menu--small mini-menu--transparent"
      >
        <button
          aria-label="Cancel"
          class="mini-menu-item--cancel mini-menu__item mini-menu__item--small"
          data-cy="undefined-item-Cancel"
          data-tooltip-content="Cancel"
          data-tooltip-id="scrumlr-tooltip"
          id="mini-menu-Cancel3"
        >
          <svg>
            close.svg
          </svg>
        </button>
        <button
          aria-label="Save"
          class="mini-menu-item--save mini-menu__item mini-menu__item--small"
          data-cy="undefined-item-Save"
          data-tooltip-content="Save"
          data-tooltip-id="scrumlr-tooltip"
          id="mini-menu-Save4"
        >
          <svg>
            check-done.svg
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ColumnConfiguratorColumnNameDetails render should render correctly (visualFeedback) 1`] = `
<div>
  <div
    class="column-configurator-column-name-details column-configurator-column-name-details__name-wrapper"
  >
    <input
      autocomplete="off"
      class="column-configurator-column-name-details__name"
      placeholder="Title"
      value="Default Name"
    />
    <div
      class="column-configurator-column-name-details__inline-description column-configurator-column-name-details__inline-description--visual-feedback"
      role="button"
      tabindex="0"
    >
      Default Description
    </div>
  </div>
</div>
`;
